pkg/mailboxwatch/impl.go

```diff
@@ -90,42 +125,104 @@ type myList struct {
+// and starts with that one next time.
+func (clw *crossClusterListerWatcher[Scoped, ListType]) List(options metav1.ListOptions) (runtime.Object, error) {
+	logger := klog.FromContext(clw.ctx)
+	allItems := []runtime.Object{}
+	var firstListRV *int64
+	var maxRV int64 = -1
+	var clusterOfMaxRV logicalcluster.Name
+	listCluster := func(clusterName logicalcluster.Name, lwForCluster *lwPerCluster[Scoped, ListType]) {
+		logger := logger.WithValues("cluster", clusterName)
+		sublist, err := lwForCluster.scopedLW.List(clw.ctx, options)
+		if err != nil {
+			if k8sapierrors.IsNotFound(err) {
+				logger.V(4).Info("Resourece not (yet) known")
+			} else {
+				logger.Error(err, "Failed to list")
+			}
+			return
+		}
+		sublistGVK := sublist.GetObjectKind().GroupVersionKind()
+		if sublistGVK != clw.listGVK && sublistGVK != (machschema.GroupVersionKind{}) {
+			logger.Error(nil, "List returned unexpected GroupVersionKind", "expected", clw.listGVK, "got", sublistGVK)
+			return
+		}
+		subItems, err := machmeta.ExtractList(sublist)
+		if err != nil {
+			logger.Error(err, "Failed to machmeta.ExtractList", "sublist", sublist)
+			return
+		}
+		listMeta, err := machmeta.ListAccessor(sublist)
+		if err != nil {
+			logger.Error(err, "Failed to machmeta.ListAccessor", "sublist", sublist)
+			return
+		}
+		rv, err := strconv.ParseInt(listMeta.GetResourceVersion(), 10, 64)
+		if err != nil {
+			logger.Error(err, "Failed to parse ResourceVersion of a List result")
+			return
+		}
+		if rv > maxRV {
+			maxRV = rv
+			clusterOfMaxRV = clusterName
+		}
+		if firstListRV == nil {
+			firstListRV = &rv
+			allItems = append(allItems, subItems...)
+			return
+		}
+		for _, item := range subItems {
+			itemM := item.(metav1.Object)
+			itemRV, err := strconv.ParseInt(itemM.GetResourceVersion(), 10, 64)
+			if err != nil {
+				logger.Error(err, "Failed to parse ResourceVersion of item", "item", item)
+				continue
+			}
+			if itemRV <= *firstListRV {
+				allItems = append(allItems, item)
+			}
+		}
+		if listMeta.GetContinue() != "" {
+			logger.Info("Warning: skipping list CONTINUE due to incomplete implementation", "gvk", clw.listGVK) // TODO: better
+		}
 	}
-	unfilteredList, err := machmeta.ExtractList(unfilteredListObj)
-	if err != nil {
-		return nil, fmt.Errorf("inner Lister produced a %T (%+v), which machmeta.ExtractList rejects: %w", unfilteredListObj, unfilteredListObj, err)
+	clw.Lock()
+	defer clw.Unlock()
+	if clw.clusterListFirst != "" {
+		if lwFirst, have := clw.perCluster[clw.clusterListFirst]; have {
+			listCluster(clw.clusterListFirst, lwFirst)
+		}
 	}
-	flw.Lock()
-	defer flw.Unlock()
-	filteredSlice := []runtime.Object{}
-	for _, item := range unfilteredList {
-		objm := item.(metav1.Object)
-		cluster := logicalcluster.From(objm)
-		if flw.clusters.Has(cluster.String()) {
-			filteredSlice = append(filteredSlice, item)
+	for clusterName, lwForCluster := range clw.perCluster {
```

Comment:
In order to make the informer work across all mailbox workspaces, CrossClusterListerWatcher is implemented so that List() returns all items in all mailbox workspaces and Watch() return watch interface that notifies all events for given GVK in all mailbox workspaces. I didn't know this useful way. I used to go with creating informer per mailbox workspace. Thanks. Will reuse this way.

Reply:
> In order to make the informer work across all mailbox workspaces, CrossClusterListerWatcher is implemented so that List() returns all items in all mailbox workspaces and Watch() return watch interface that notifies all events for given GVK in all mailbox workspaces.
> Is my understanding correct?

I would say "yes".

>  If so, I didn't know this useful way. I used to go with creating informer per mailbox workspace. Thanks. (strirkethrough)Will reuse this way in C2P.(/strikethrough) => once it's merged, no need any treatment in the consumer of the informer library.

I am not sure what you mean.  Could you maybe rephrase or expand on that?

Reply:
To make informer support non-builtin API across workspaces,  I was implementing the informer with naive approach that creates informer per workspace as commented on https://github.com/kcp-dev/edge-mc/pull/265#issuecomment-1495581481. I couldn't have found an alternative approach rather than the naive approach.  This approach is very helpful that I do not have to create as many informers as mailbox workspace. In that sense, I would say thank you that I could know that technique. 
As for the last scratched comment, I initially thought that I have to implement listerWatcher to have List() and Watch() to return items and watcher aggregated over workspaces following to your code for my purpose (collecting upsynced reports) but I realized it was wrong. I don't need to implement listerWatcher but just OK to put an object that implements ClusterInterface (e.g. `edgescopedclient.SyncerConfigInterface`, `dynamic.ClusterClientset`) to `mailboxwatch.NewSharedInformer`. `mailboxwatch.NewSharedInformer` does wrap it with newCrossClusterListerWatcher.

Reply:
@yana1205 : I am still not sure I fully understand.  If you are saying that you can replace your client-side collection of informers with one mailboxwatch informer now --- due to the changes in this PR --- then I agree.  The C2P controller should only need to make one call, like https://github.com/MikeSpreitzer/kcp-edge-mc/blob/f0c0c9ed8a6d1511256b84a51882797dc90e7607/cmd/inform-syncer-configs/main.go#L93 , for each cluster-scoped resource, passing a listerwatcher directly supplied by the generated client code such as in https://github.com/MikeSpreitzer/kcp-edge-mc/blob/f0c0c9ed8a6d1511256b84a51882797dc90e7607/cmd/inform-syncer-configs/main.go#L83 .  For resources the are namespace scoped rather than cluster scoped, you need to use https://github.com/MikeSpreitzer/kcp-edge-mc/blob/f0c0c9ed8a6d1511256b84a51882797dc90e7607/pkg/mailboxwatch/listwatch.go#L71-L77 to convert the listerwatcher from the client code to make what NewSharedInformer needs.

Reply:
> If you are saying that you can replace your client-side collection of informers with one mailboxwatch informer now
> The C2P controller should only need to make one call,

Yes, that's what I wanted to say.

> For resources the are namespace scoped rather than cluster scoped, you need to use https://github.com/MikeSpreitzer/kcp-edge-mc/blob/f0c0c9ed8a6d1511256b84a51882797dc90e7607/pkg/mailboxwatch/listwatch.go#L71-L77 to convert the listerwatcher from the client code to make what NewSharedInformer needs.

I could invoke the informer for namespaced resources too following the guide above and got the namespaced objects in all mailbox workspaces. Thanks.

/lgtm

Reply:
@yana1205 : I am not sure I understand what you are saying about namespaced resources.  Can you show or refer me to the code?

Reply:
Here is the snippet.
```
	gvk := schema.GroupVersionKind{
		Group:   gvr.Group,
		Version: gvr.Version,
		Kind:    "ConfigMapList",
	}

	// use mailboxwatch.FixNamespace
	kcpClientSet, err := kcpkubernetes.NewForConfig(allClientConfig)
	if err != nil {
		logger.Error(err, "Failed to create all-cluster dynamic client")
		os.Exit(30)
	}
	configmapSets := kcpClientSet.CoreV1().ConfigMaps()
	lw := mailboxwatch.FixNamespace[corev1.ConfigMapsNamespacer, k8scoresv1.ConfigMapInterface, *coreapi.ConfigMapList](configmapSets, metav1.NamespaceAll)
	informer1 := mailboxwatch.NewSharedInformer(ctx, gvk, mbPreInformer, lw, &coreapi.ConfigMap{}, 0, upstreamcache.Indexers{})
	informer1.AddEventHandler(upstreamcache.ResourceEventHandlerFuncs{
		AddFunc:    func(obj any) { logCm(logger, "add 1", obj) },
		UpdateFunc: func(oldObj, newObj any) { logCm(logger, "update 1", newObj) },
		DeleteFunc: func(obj any) { logCm(logger, "delete 1", obj) },
	})
```

Here is the entire code: https://gist.github.com/yana1205/40ce9e8fe353164dd9ec65b787bbc4a2

e.g.
```
$ KUBECONFIG=/tmp/admin.kubeconfig go run cmd/inform-namespaced-resources/main.go
I0420 13:51:06.074886   30369 main.go:136] "Running"
I0420 13:51:06.078322   30369 main.go:148] "Notified" action="add 1" name="kyverno" namespace="kyverno" cluster="2c2gizqmvm4vlt81"
I0420 13:51:06.078414   30369 main.go:148] "Notified" action="add 1" name="kube-root-ca.crt" namespace="policy" cluster="2c2gizqmvm4vlt81"
I0420 13:51:06.078426   30369 main.go:148] "Notified" action="add 1" name="kube-root-ca.crt" namespace="default" cluster="2c2gizqmvm4vlt81"
I0420 13:51:06.078435   30369 main.go:148] "Notified" action="add 1" name="kube-root-ca.crt" namespace="kyverno" cluster="2c2gizqmvm4vlt81"
I0420 13:51:06.078443   30369 main.go:148] "Notified" action="add 1" name="kyverno-metrics" namespace="kyverno" cluster="2c2gizqmvm4vlt81"
I0420 13:51:06.078476   30369 main.go:148] "Notified" action="add 1" name="kyverno-metrics" namespace="kyverno" cluster="rqkzl5vi2uzrj3ec"
I0420 13:51:06.078494   30369 main.go:148] "Notified" action="add 1" name="kyverno" namespace="kyverno" cluster="rqkzl5vi2uzrj3ec"
I0420 13:51:06.078503   30369 main.go:148] "Notified" action="add 1" name="kube-root-ca.crt" namespace="policy" cluster="rqkzl5vi2uzrj3ec"
I0420 13:51:06.078512   30369 main.go:148] "Notified" action="add 1" name="kube-root-ca.crt" namespace="default" cluster="rqkzl5vi2uzrj3ec"
I0420 13:51:06.078521   30369 main.go:148] "Notified" action="add 1" name="kube-root-ca.crt" namespace="kyverno" cluster="rqkzl5vi2uzrj3ec"
I0420 13:51:06.078755   30369 main.go:156] "Notified" action="add 2" name="kube-root-ca.crt" namespace="default" cluster="2c2gizqmvm4vlt81"
I0420 13:51:06.078771   30369 main.go:156] "Notified" action="add 2" name="kube-root-ca.crt" namespace="kyverno" cluster="2c2gizqmvm4vlt81"
I0420 13:51:06.078779   30369 main.go:156] "Notified" action="add 2" name="kyverno-metrics" namespace="kyverno" cluster="2c2gizqmvm4vlt81"
I0420 13:51:06.078783   30369 main.go:156] "Notified" action="add 2" name="kyverno" namespace="kyverno" cluster="2c2gizqmvm4vlt81"
I0420 13:51:06.078788   30369 main.go:156] "Notified" action="add 2" name="kube-root-ca.crt" namespace="policy" cluster="2c2gizqmvm4vlt81"
I0420 13:51:06.078792   30369 main.go:156] "Notified" action="add 2" name="kyverno-metrics" namespace="kyverno" cluster="rqkzl5vi2uzrj3ec"
I0420 13:51:06.078797   30369 main.go:156] "Notified" action="add 2" name="kyverno" namespace="kyverno" cluster="rqkzl5vi2uzrj3ec"
I0420 13:51:06.078802   30369 main.go:156] "Notified" action="add 2" name="kube-root-ca.crt" namespace="policy" cluster="rqkzl5vi2uzrj3ec"
I0420 13:51:06.078908   30369 main.go:156] "Notified" action="add 2" name="kube-root-ca.crt" namespace="kyverno" cluster="rqkzl5vi2uzrj3ec"
I0420 13:51:06.078925   30369 main.go:156] "Notified" action="add 2" name="kube-root-ca.crt" namespace="default" cluster="rqkzl5vi2uzrj3ec"
```

Reply:
Sorry, the above one contains another informer who is initialized with dynamicClient.
I removed that part to avoid confusion. Here is the updated one: https://gist.github.com/yana1205/061ae7b6623e8ed4c024b60ae3eec9fd

```
$ KUBECONFIG=/tmp/admin.kubeconfig go run cmd/inform-namespaced-resources/main.go
I0420 13:59:00.847118   31952 main.go:122] "Running"
I0420 13:59:00.848989   31952 main.go:134] "Notified" action="add 1" name="kyverno" namespace="kyverno" cluster="2c2gizqmvm4vlt81"
I0420 13:59:00.849006   31952 main.go:134] "Notified" action="add 1" name="kube-root-ca.crt" namespace="policy" cluster="2c2gizqmvm4vlt81"
I0420 13:59:00.849011   31952 main.go:134] "Notified" action="add 1" name="kyverno-metrics" namespace="kyverno" cluster="2c2gizqmvm4vlt81"
I0420 13:59:00.849014   31952 main.go:134] "Notified" action="add 1" name="kube-root-ca.crt" namespace="default" cluster="2c2gizqmvm4vlt81"
I0420 13:59:00.849018   31952 main.go:134] "Notified" action="add 1" name="kube-root-ca.crt" namespace="kyverno" cluster="2c2gizqmvm4vlt81"
I0420 13:59:00.849022   31952 main.go:134] "Notified" action="add 1" name="kyverno-metrics" namespace="kyverno" cluster="rqkzl5vi2uzrj3ec"
I0420 13:59:00.849026   31952 main.go:134] "Notified" action="add 1" name="kyverno" namespace="kyverno" cluster="rqkzl5vi2uzrj3ec"
I0420 13:59:00.849029   31952 main.go:134] "Notified" action="add 1" name="kube-root-ca.crt" namespace="policy" cluster="rqkzl5vi2uzrj3ec"
I0420 13:59:00.849033   31952 main.go:134] "Notified" action="add 1" name="kube-root-ca.crt" namespace="kyverno" cluster="rqkzl5vi2uzrj3ec"
I0420 13:59:00.849036   31952 main.go:134] "Notified" action="add 1" name="kube-root-ca.crt" namespace="default" cluster="rqkzl5vi2uzrj3ec"
```

Reply:
> I am not sure I understand what you are saying about namespaced resources

In the case of namespaced resource like ConfigMap, I thought that users have to create mailboxwatch.FixNamespace firstly, then pass it to mailboxwatch.NewSharedInformer. I wanted to just say is that I tried and confirmed this pattern.

Reply:
OK, yes, that's right Thanks!